### PR TITLE
GH-98539: fix ref cycle in `_SSLProtocolTransport` after close 

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-98539

### CODE REVIEW

File: Lib/asyncio/sslproto.py [hunk 1]
```diff
@@ -107,8 +107,11 @@ def close(self):
         protocol's connection_lost() method will (eventually) called
         with None as its argument.
         """
-        self._closed = True
-        self._ssl_protocol._start_shutdown()
+        if not self._closed:
+            self._closed = True
+            self._ssl_protocol._start_shutdown()
```
Comment (gvanrossum) [inline Lib/asyncio/sslproto.py:112]: So if we're taking this path through the code, the cycle remains unbroken?

Comment (kumaraditya303) [reply Lib/asyncio/sslproto.py:112]: The cycle will be broken when the `_ssl_protocol` calls this `close` method again. When the shutdown is complete this method is called again and in that case `_closed` is true so we break the cycle.

Comment (kumaraditya303) [reply Lib/asyncio/sslproto.py:112]: See https://github.com/python/cpython/blob/5871e19942fdcf83653924ae9a849941669c1143/Lib/asyncio/sslproto.py#L656-L664

The last line schedules to call this again when the whole shutdown is complete.

Comment (gvanrossum) [reply Lib/asyncio/sslproto.py:112]: How bizarre that it relies on calling `close()` twice. But everything about SSL is bizarre, so I'll merge this now. :-)

**Review by gvanrossum — APPROVED** [review]

File: Lib/asyncio/sslproto.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/sslproto.py:112]: "If it ain't broke, don't fix it" :)

**Comment by Dreamsorcerer** [comment]
This change makes it unsafe to call `.abort()` after being closed (which seems to be causing exceptions and severe memory leaks for some aiohttp users in 3.11.1+). We'll obviously add a check to aiohttp, but should `.abort()` be updated to be safe to call?

**Comment by gvanrossum** [comment]
> This change makes it unsafe to call `.abort()` after being closed (which seems to be causing exceptions and severe memory leaks for some aiohttp users in 3.11.1+). We'll obviously add a check to aiohttp, but should `.abort()` be updated to be safe to call?

Makes sense. Can you send a PR?

**Comment by Dreamsorcerer** [comment]
@gvanrossum #104474

**Pull request merged**